test/e2e/suite/nodeadm_test.go

```diff
@@ -348,12 +291,204 @@ var _ = Describe("Hybrid Nodes", func() {
+								os:                 os,
+								provider:           provider,
+								s3Client:           test.s3Client,
+								setRootPassword:    test.setRootPassword,
+								ssmClient:          test.ssmClient,
+							}
+							instance, err := createNodeTest.Run(ctx)
+							Expect(err).NotTo(HaveOccurred(), "EC2 Instance should have been created successfully")
+
+							joinNodeTest := joinNodeTest{
+								k8s:           test.k8sClient,
+								nodeIPAddress: instance.IP,
+								logger:        test.logger,
+							}
+							Expect(joinNodeTest.Run(ctx)).To(Succeed(), "node should have joined the cluster sucessfully")
+
+							upgradeNodeTest := upgradeNodeTest{
+								k8s:      test.k8sClient,
+								ssm:      test.ssmClient,
+								cluster:  test.cluster,
+								ec2:      instance,
+								logger:   test.logger,
+								provider: provider,
+							}
+							Expect(upgradeNodeTest.Run(ctx)).To(Succeed(), "node should have upgraded successfully")
+							Expect(joinNodeTest.Run(ctx)).To(Succeed(), "node should have joined the cluster sucessfully after nodeadm upgrade")
+
+							test.logger.Info("Resetting hybrid node...")
+
+							uninstallNodeTest := uninstallNodeTest{
+								k8s:      test.k8sClient,
+								ssm:      test.ssmClient,
+								ec2:      instance,
+								provider: provider,
+								logger:   test.logger,
+							}
+							Expect(uninstallNodeTest.Run(ctx)).To(Succeed(), "node should have been reset sucessfully")
+						},
+						Entry(fmt.Sprintf("With OS %s and with Credential Provider %s", os.Name(), string(provider.Name())), context.Background(), os, provider, Label(os.Name(), string(provider.Name()), "upgradeflow")),
+					)
 				}
 			}
 		})
 	})
 })
 
+type createNodeTest struct {
+	awsSession         *session.Session
+	cluster            *peered.HybridCluster
+	ec2ClientV2        *ec2v2.Client
+	instanceName       string
+	instanceProfileARN string
+	k8sClient          *clientgo.Clientset
+	k8sVersion         string
+	logger             logr.Logger
+	logsBucket         string
+	nodeadmURLs        e2e.NodeadmURLs
+	os                 e2e.NodeadmOS
+	provider           e2e.NodeadmCredentialsProvider
+	s3Client           *s3v1.S3
+	setRootPassword    bool
+	skipCleanup        bool
+	ssmClient          *ssmv1.SSM
+}
+
+func (c createNodeTest) Run(ctx context.Context) (ec2.Instance, error) {
+
+	nodeSpec := e2e.NodeSpec{
+		OS: c.os,
+		Cluster: &e2e.Cluster{
+			Name:   c.cluster.Name,
+			Region: c.cluster.Region,
+		},
+		Provider: c.provider,
+	}
+
+	files, err := c.provider.FilesForNode(nodeSpec)
+	if err != nil {
+		return ec2.Instance{}, err
+	}
+
+	nodeadmConfig, err := c.provider.NodeadmConfig(ctx, nodeSpec)
+	if err != nil {
+		return ec2.Instance{}, fmt.Errorf("expected to build nodeconfig: %v", err)
+	}
+
+	nodeadmConfigYaml, err := yaml.Marshal(&nodeadmConfig)
+	if err != nil {
+		return ec2.Instance{}, fmt.Errorf("expected to successfully marshal nodeadm config to YAML: %v", err)
```

Comment:
prefer to use %w instead of %v for errors. %w wraps the error, so things like `Unwrap`/`As`/`Is` work properly
Sometimes you might not want to wrap, like if you want to hide the underlying type and not allow to check for it up the chain. But in most cases wrap is what you want

```suggestion
		return ec2.Instance{}, fmt.Errorf("expected to successfully marshal nodeadm config to YAML: %w", err)
```